src/demangler/demangler.cc

```diff
@@ -0,0 +1,155 @@
+  for (; ptr <= end; ++ptr) {
+    if (*ptr == '$') {
+      if (ptr[1] == '$') {
+        return false;
+      } else if (ptr[2] == '$' || ptr[3] == '$' || ptr[4] == '$') {
+        return true;
+      } else {
+        return false;
+      }
+    } else if (*ptr == '.') {
+      return !('.' == ptr[1] &&
+               '.' == ptr[2]); // '.' and '..' are fine, '...' is not
+    }
+  }
+  return true;
+}
+
+// Simple conversion from hex digit to integer
+// Includes capitals out of completeness, but this should not be necessary for
+// the implementation
+inline static int hex_to_int(char dig) {
+  if (dig >= '0' && dig <= '9')
+    return dig - '0';
+  else if (dig >= 'a' && dig <= 'f')
+    return dig - 'a' + 10;
+  else if (dig >= 'A' && dig <= 'F')
+    return dig - 'A' + 10;
+  else
+    return -1;
+}
+
+// Demangles a Rust string by building a copy piece-by-piece
+static std::string rust_demangle(const std::string &str) {
+  static const std::unordered_map<std::string, std::string> patterns{
+      {"..", "::"},  {"$C$", ","},  {"$BP$", "*"}, {"$GT$", ">"}, {"$LT$", "<"},
+      {"$LP$", "("}, {"$RP$", ")"}, {"$RF$", "&"}, {"$SP$", "@"},
+  };
+
+  std::string ret;
+  ret.reserve(str.size());
+
+  size_t i = 0;
+
+  // Special-case for repairing C++ demangling defect for Rust
+  if (str[0] == '_' && str[1] == '$')
+    ++i;
+
+  for (; i < str.size() - hash_pre.size() - hash_eg.size(); ++i) {
+
+    // Fast sieve for pattern-matching, since we know first chars
+    if (str[i] == '.' || str[i] == '$') {
+      bool replaced = false;
+
+      // Try to replace one of the patterns
+      for (const auto &pair : patterns) {
+        const std::string &pattern = pair.first;
+        const std::string &replacement = pair.second;
+        if (!str.compare(i, pattern.size(), pattern)) {
+          ret += replacement;
+          i += pattern.size() - 1; // -1 because iterator inc
+          replaced = true;
+          break;
+        }
+      }
+
+      // If we failed to replace, try a few failovers.  Notably, we recognize
+      // that Rust may insert Unicode code points in the function name (other
+      // implementations treat many individual points as patterns to search on)
+      if (!replaced && str[i] == '.') {
+        // Special-case for '.'
+        ret += '-';
+      } else if (!replaced && !str.compare(i, 2, "$u") && str[i + 4] == '$') {
+        int hi = hex_to_int(str[i + 2]);
+        int lo = hex_to_int(str[i + 3]);
+        if (hi != -1 && lo != -1) {
+          ret += static_cast<unsigned char>(lo + 16 * hi);
+          i += 5 - 1; // - 1 because iterator inc
+        } else {
+          // We didn't have valid unicode values, but we should still skip
+          // the $u??$ sequence
+          ret += str.substr(i, 5);
+          i += 5 - 1; // -1 because iterator inc
+        }
+      } else if (!replaced) {
+        ret += str[i];
+      }
+    } else {
+      ret += str[i];
+    }
+  }
+
+  ret.shrink_to_fit();
```

Comment:
Not sure this is useful, in most cases demangled string will be larger than source string, and in the cases where it's not, losing a few bytes of memory should not be an issue.

Reply:
I agree that this probably isn't _too_ impactful--on the order of ~19 bytes per string, plus a little bit more for strings with a lot of pattern substitutions.

At this point in processing, the input string has undergone C++ demangling (making it bigger than the source string).  This function acts on the resulting string by
1. Dropping a terminating hash (16 + 3 characters)
2. Converting patterns of length 3-5 into single characters (or in one case, two characters)

I guess most of that space could be restored by just not reserving the full size of the input on line 93, but rather `str.size() - hash_eg.size() - hash_pre.size())` or something, which is probably better than shrinking at the end.  In that case, I don't think the few bytes we gain by shrinking in the average case are worth it (ddprof manages its own deduplication for function names, but when strings are transferred to libdatadog I anticipate them to be copied by size rather than capacity).

I'll modify this accordingly--thanks for noticing

Reply:
My bad, please disregard my comment.
I completely missed the fact that the hash at the end was removed !